fix(platform-browser): only one hammer instance is created in one element

sculove opened the pull request:
Closes #16333

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[X] Other... Please describe: I added hammerjs in karma.conf
```

**What is the current behavior?** (You can also link to an open issue here)
#16333


**What is the new behavior?**
When Hammer instance is created I check it.
If instance is created, reuse it!


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[X] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

splincode commented:
@sculove what is the status of your pull request? is it still relevant? please fix conflicts if they exist

trotyl committed:
fix(core): support computed base class in metadata inheritance (#24014)

PR Close #24014

LayZeeDK committed:
feat(forms): add updateOn option to FormBuilder (#24599)

PR Close #24599

sjtrimble committed:
build(docs-infra): improve API overload templates (#24976)

PR Close #24976

petebacondarwin committed:
build(docs-infra): add method overload index (#24976)

PR Close #24976

petebacondarwin committed:
build(docs-infra): fix individual API overload templates (#24976)

* Make individual overloads collapsible
* Show only the first overload expanded, rest collapsed
* Text changes to 'collapse all' once 'show all' is clicked
* Fix chevron/carrot rotation animation when overloads / overload item is expanded or collapsed

PR Close #24976

petebacondarwin committed:
build(docs-infra): display long overload parameter types as `object` (#24976)

In some overloads, the parameter type can be a large anonymous
object type.
This change displays such types as `object`. It is then up to the
documentation author to put more information about the type in the
method usage notes.

PR Close #24976

joostme committed:
feat(service-worker): handle 'notificationclick' events (#25860)

The previous version did not support the 'notificationclick' event.
Add event handler for the event and provide an observable of
clicked notifications in the SwPush service.

Closes #20956, #22311

PR Close #25860

joostme committed:
fix(service-worker): add missing api typing (#25860)

Add missing messagesClicked typing for SwPush service

PR Close #25860

shinok01 committed:
fix(core): ignore comment nodes under unsafe elements (#25879)

Comment nodes that are child nodes of unsafe elements are identified as text nodes. This results in the comment node being returned as an encoded string.
Add a check to ignore such comment nodes.

PR Close #25879

joostme committed:
refactor(service-worker): update tests based of review feedback (#25860)

- Rename invalid click event name
- Make test expects more explicit
- Remove NotificationClickEvent in favor of NotificationEvent

PR Close #25860

joostme committed:
feat(service-worker): close notifications and focus window on click (#25860)

- Serialize notification object before using postMessage
- Close notification on click
- Focus browser if it is not already focused on click

PR Close #25860

alexeagle committed:
build: introduce a package.bzl (#26488)

This lets Angular Bazel users install our transitive deps, rather than have to list them in their WORKSPACE file.
If they want a different version of one of these deps, they just need to install it before calling rules_angular_dependencies.

PR Close #26488

joostme committed:
refactor(service-worker): Code optimizations (#25860)

- Add missing new line in tests
- Add proper noop method in MockNotificationEvent

PR Close #25860

joostme committed:
feat(service-worker): Add typing for messagesClicked in SwPush service (#25860)

- Properly type messagesClicked Observable stream
- Add new NotificationObject interface

PR Close #25860

joostme committed:
refactor(service-worker): Rework notification click handler (#25860)

- Add missing image and timestamp properties
- Remove focus from click handler

PR Close #25860

joostme committed:
fix(service-worker): Add typing to public api guard and fix lint errors (#25860)

PR Close #25860

joostme committed:
fix(service-worker): Fix public api guard typing (#25860)

PR Close #25860

joostme committed:
refactor(service-worker): Format comments and add additional test (#25860)

- Format JSDoc for notificationClicks
- Add comment on why handleClick does not use hasOwnProperty
- Add additional test that uses handleClick without action

PR Close #25860

jasonaden committed:
refactor(router): add type guards for router guards (#26521)

PR Close #26521

jasonaden committed:
refactor(router): add type definitions for guard functions (#26521)

PR Close #26521

gregmagolan committed:
build(bazel): use fine-grained npm deps (#26111) (#26488)

PR Close #26488

petebacondarwin committed:
fix(core): ensure that `ɵdefineNgModule` is available in flat-file formats (#26403)

When compiling the flat-file version of the `@angular/core` we need to be aware
that we cannot rely upon imported names to access the ivy definition functions.
The compiler is already clever enough to use local function calls rather than
trying to add a namespaced import, but there is a problem if the local name of the
function is different to the exported name. This is the case for functions that
are not part of the public API, and so are exported under a barred-O private alias.

In `@angular/core` the only decorations in use are `@NgModule` and `@Injectable`.
There are no directives, components, pipes, etc.

Since `defineInjectable` is part of the public API of `@angular/core`, the compiler
is able to generate code that references the original non-barred-O version of the
function.

But the `defineNgModule` is not part of the public API and so the compiler must
generate code that refers to it by the private barred-O version of the function.

This commit imports and then re-exports this barred-O version of `defineModule` to
ensure that the symbol is available in the local scope of the flat-file versions of
the `@angular/core` library.

PR Close #26403

jasonaden committed:
refactor(router): remove type casting from wrapIntoObservable private function (#26521)

PR Close #26521

petebacondarwin committed:
build(docs-infra): render CLI arguments consistently (#26272)

In the command syntax, arguments are rendered as
`var`s enclosed in angle brackets. So this is now repeated
in the arguments table too.

PR Close #26272

petebacondarwin committed:
build(docs-infra): update CLI option rendering (#26272)

PR Close #26272

petebacondarwin committed:
build(docs-infra): break long CLI options onto two lines (#26272)

PR Close #26272

gregmagolan committed:
build: update to rules_typescript 0.20.1 and rules_nodejs 0.15.0 (#26260) (#26488)


PR Close #26488

petebacondarwin committed:
test(ivy): ngcc - test compiling a CLI generated project (#26403)

This integration test was created from a vanilla CLI generated
project with the following modifications:

* remove `PercentPipe` usage from `app.component.html`
  - these are not yet supported by ivy
* changed `ng test` in `package.json` to only to `ng build`
  - right now we can only confirm that the app will build
* hard-code `ngDevMode` in `index.html`
  - the CLI does not yet set this correctly

PR Close #26403

jasonaden committed:
feat(router): allow redirect from guards by returning UrlTree (#26521)

* Improve type checking within the CheckGuards function
* Change public API to allow returning of UrlTree instead of boolean

PR Close #26521

gregmagolan committed:
build(bazel): update to rules_typescript 0.20.2 (#26279) (#26488)


PR Close #26488

petebacondarwin committed:
refactor(ivy): ngcc - recombine flat and non-flat `Esm2015ReflectionHost` (#26403)

Going forward we need to be able to do the same work on both
flat and non-flat module formats (such as computing arity and
transforming .d.ts files)

PR Close #26403

petebacondarwin committed:
fix(ivy): ngcc - support Angular Material package.json format (#26403)

The Material project uses slightly different properties to the
core Angular project for specifying the different format entry-point.

This commit ensures that we map these properties correctly for both
types of project.

PR Close #26403

gkalpak committed:
test(docs-infra): improve logging output in `test-pwa-score[-localhost]` (#26377)

PR Close #26377

petebacondarwin committed:
refactor(ivy): ngcc - move typings rendering to `Renderer` (#26403)

The rendering of typings is not specific to the package
format, so it doesn't make sense to put it in a specific
renderer.

As a result there is no real difference between esm5 and esm2015
renderers, so there is no point in having separate classes.

PR Close #26403

gkalpak committed:
ci(docs-infra): add jobs to CircleCI config (#26377)

PR Close #26377

lovelangy committed:
docs: type fix (#26386)

according to the source code and the doc https://angular.io/api/forms/NgModel#inherited-from-formsngcontrol-1, the method name should be "viewToModelUpdate" instead of "viewToModel"
PR Close #26386

nik72619c committed:
docs(service-worker): updated browser support for Service Worker (#26408)

PR Close #26408

mhevery commented:
http://test/OCL:216779398:BASE:216779438:1539299785000:dd0bb8ec

sculove commented:
I will resend PR. Thanks

mhevery commented:
> I will resend PR. Thanks

So does that means I should not merge the PR?

sculove commented:
I think you'd better merge this pr.
Thanks.

jenniferfell committed:
docs: update npm packages to be accurate for v7 (#26422)

PR Close #26422

jbogarthyde committed:
docs: configuration file reference (#26484)

PR Close #26484

petebacondarwin committed:
test(ivy): ngcc - test compiling the Angular Material library (#26403)

* rename test helper script
* add material to the ngcc integration test
* add MatButton to ngcc integration test checks
* remove platform-server from ngcc integration test
  This package does not yet compile as it contains a package-private
  (internal) decorated class, which the ngcc compiler does not yet
  handle.

PR Close #26403

petebacondarwin committed:
feat(ivy): support separate .js and .d.ts trees when generating imports (#26403)

The `NgModule` handler generates `R3References` for its declarations, imports,
exports, and bootstrap components, based on the relative import path
between the module and the classes it's referring to. This works fine for
compilation of a .ts Program inside ngtsc, but in ngcc the import needed
in the .d.ts file may be very different to the import needed between .js
files (for example, if the .js files are flattened and the .d.ts is not).

This commit introduces a new API in the `ReflectionHost` for extracting the
.d.ts version of a declaration, and makes use of it in the
`NgModuleDecorationHandler` to write a correct expression for the `NgModule`
definition type.

PR Close #26403

petebacondarwin committed:
fix(ivy): fix 'Module not found` error message (#26403)

The message was dumping a serialized object instead of a
human readable name into the exception being thrown.

PR Close #26403

brandonroberts committed:
docs: forms overview copy edit (#26450)

PR Close #26450

petebacondarwin committed:
fix(ivy): ngcc - skip missing formats rather than erroring (#26403)

It is perfectly normal for some of the formats to be missing from
a package. We should not fail compilation for this.

PR Close #26403

petebacondarwin committed:
perf(ivy): ngcc - use flat file for dependency sorting if available (#26403)

Previously we always used the non-flat format because we thought
that this was the one that would always be available.

It turns out that this is not the case and that only one of the flat and
non-flat formats may be available.

Therefore we should use whichever is available, defaulting to the flat
format if that exists, since that will be faster to parse.

PR Close #26403

alexeagle committed:
build: update ngcontainer to bazel 0.18.0 (#26465) (#26488)

* build: update ngcontainer to bazel 0.18.0

* build: update skylint to bazel 0.18

use .bazelignore file to ignore node_modules directory

PR Close #26488

Nazanin1369 committed:
docs(service-worker): Specify format of datagroups.cacheConfig.timeout (#26469)

Fixes #26454

PR Close #26469

IgorMinar committed:
test(ivy): mark failing test targets with fixme-ivy-jit and fixme-ivy-local tags (#26471)

We are close enough to blacklist a few test targets, rather than whitelist targets to run...

Because bazel rules can be composed of other rules that don't inherit tags automatically,
I had to explicitly mark all of our ts_library and ng_module targes with "ivy-local" and
"ivy-jit" tags so that we can create a query that excludes all fixme- tagged targets even
if those targets are composed of other targets that don't inherit this tag.

This is the updated overview of ivy related bazel tags:

- ivy-only: target that builds or runs only under ivy
- fixme-ivy-jit: target that doesn't yet build or run under ivy with --compile=jit
- fixme-ivy-local: target that doesn't yet build or run under ivy with --compile=local
- no-ivy-jit: target that is not intended to build or run under ivy with --compile=jit
- no-ivy-local: target that is not intended to build or run under ivy with --compile=local

PR Close #26471

petebacondarwin committed:
fix(ivy): ngcc - fixes to support compiling Material library (#26403)

1) The `DecorationAnalyzer now analyzes all source files, rather than just
the entry-point files, which fixes #26183.
2) The `DecoratorAnalyzer` now runs all the `handler.analyze()`  calls
across the whole entry-point *before* running `handler.compile()`. This
ensures that dependencies between the decorated classes *within* an
entry-point are known to the handlers when running the compile process.
3) The `Renderer` now does the transformation of the typings (.d.ts) files
which allows us to support packages that only have flat format
entry-points better, and is faster, since we won't parse `.d.ts` files twice.

PR Close #26403

petebacondarwin committed:
perf(ivy): ngcc - only render .d.ts analysis when necessary (#26403)

For each package entry-point there is only one format that
is used to compile the typings files (.d.ts). This will be
either esm2015 or fesm2015 (preferred). So we would not run
any dts processing in the renderer if we are not compiling
the appropriate format.

PR Close #26403

gkalpak committed:
test(animations): make e2e tests for animations docs examples less flaky (#26377)

PR Close #26377

gkalpak committed:
test(elements): make e2e tests for elements docs examples less flaky (#26377)

PR Close #26377

AndrewKushnir committed:
fix(ivy): move HostListeners generation to factory function (#26480)

PR Close #26480

alxhub committed:
refactor(ivy): rename ngtsc/factories to ngtsc/shims (#26495)

This simple refactor of the build rules renames the .ngfactory.js shim
generator to 'shims' instead of 'factories', in preparation for adding
.ngsummary.js shim generation.

Testing strategy: this commit does not introduce any new behavior and
merely moves files and symbols around. It's sufficient that the existing
ngtsc tests pass.

PR Close #26495

null-reference committed:
docs: removing errant double-quote (#26483)
PR Close #26483

IgorMinar committed:
docs: fixup deployment guide (#26486)

- remove sections that were not relevant (all the development stuff)
- fix incorrect info

PR Close #26486

gregmagolan committed:
build: idiomatic install of @angular/bazel npm package (#26258) (#26488)


PR Close #26488

gregmagolan committed:
build: restore integrity lines in yarn.lock (#26488)

PR Close #26488

gregmagolan committed:
build: add npm deps after rebase (#26488)

PR Close #26488

mhevery commented:
http://test/OCL:217389628:BASE:217389667:1539726221840:82409829

alxhub committed:
refactor(ivy): make shim generation generic in ngtsc (#26495)

This commit refactors the shim host to be agnostic to the shims being
generated, and provides an API for generating additional shims besides
the .ngfactory.js. This will be used in a following commit to generate
.ngsummary.js shims.

Testing strategy: this refactor introduces no new behavior, so it's
sufficient that the existing tests for factory shim generation continue
to pass.

PR Close #26495

alxhub committed:
feat(ivy): generate .ngsummary.js shims (#26495)

This commit adds generation of .ngsummary.js shims alongside .ngfactory.js
shims when generated files are enabled.

Generated .ngsummary shims contain a single, null export for every exported
class with decorators that exists in the original source files. Ivy code
does not depend on summaries, so these exist only as a placeholder to allow
them to be imported and their values passed to old APIs. This preserves
backwards compatibility.

Testing strategy: this commit adds a compiler test to verify the correct
shape and contents of the generated .ngsummary.js files.

PR Close #26495

gregmagolan committed:
build: update to rules_typescript 0.20.3 and rules_nodejs 0.15.1 (#26488)

PR Close #26488

jasonaden committed:
fix(router): remove type bludgeoning of context and outlet when running CanDeactivate (#26496)

Fixes #18253

PR Close #26496

jasonaden committed:
feat(router): allow guards to return UrlTree as well as boolean (#26521)

* Removed `andObservable` helper function in favor of inline implementation
* Flow `boolean | UrlTree` through guards check
* Add tests to verify behavior of `checkGuards` function flowing `UrlTree` properly

PR Close #26521

ocombe committed:
docs(ivy): update status for i18n (#26502)

PR Close #26502

pkozlowski-opensource committed:
fix(ivy): allow root components to inject ViewContainerRef (#26682)

PR Close #26682

Commit:
fix(aio): add relative to app level routes section (#26504)

Added word *relative* to the **Routes at the app level** section description. This was not specified before. The routes in the *lazy-loading-ngmodules/src/app/app-routing.module.ts* also had `loadChildren` values starting with `app/...`.

The code for `app-routing.module.ts` is already fixed in [this commit](https://github.com/angular/angular/commit/67ad9468d35fee9e0b0f829d18d01eafda251f1e)
PR Close #26504

pkozlowski-opensource committed:
fix(ivy): stub TestBed.compileComponents implementation (#26506)

PR Close #26506

petebacondarwin committed:
build(docs-infra): show github edit link on CLI overview (#26515)

PR Close #26515

alan-agius4 committed:
docs: replace `alert-is-helpful` with `alert is-helpful` (#26497)

PR Close #26497

alan-agius4 committed:
docs: update continuous integration story (#26497)

PR Close #26497

alan-agius4 committed:
docs: update contributing page (#26497)

PR Close #26497

robphoenix committed:
docs: Update link to angular-cli repo (#26497)
PR Close #26497

robphoenix committed:
docs: fix formatting of PULL_REQUEST_TEMPLATE.md (#26512)

This change enables developers to check the checkboxes or update them
after they submitted a PR.

PR Close #26512

gkalpak committed:
build(docs-infra): throw error if using `title` on code snippets (#26514)

Since #26396, the `title` property is ignored and `header` should be
used instead for specifying a code snippet's header.

This commit ensures that we don't accidentally set `title` have it be
silently ignored.

PR Close #26514

petebacondarwin committed:
build(docs-infra): display github links in CLI API docs (#26515)

This commit includes the following changes:

* CLI version information is read from the CLI package from which
  we read the help files.
* CLI API pages now contain GH links
* line numbers are not shown in GH links, if the doc does not
  have a truthy `startingLine` value. This allows us to remove
  hard coded checks for `guide` pages
* content pages and CLI api docs no longer have a `startingLine`
* the hard-coded `packages` path segment has been removed from
  the templates; instead we now only use the `realProjectRelativePath`.
* the `realProjectRelativePath` has been updated accordingly for API
  and CLI API docs.

PR Close #26515

alan-agius4 committed:
build(docs-infra): update git ref for cli command docs (#26497)

Update for version 7 release

PR Close #26497

alexeagle committed:
refactor(bazel): remove workaround for TS bug (#26516)

PR Close #26516

gkalpak committed:
build: use CI-provider independent variable names (#26377)

PR Close #26377

jasonaden committed:
feat(router): guard returning UrlTree cancels current navigation and redirects (#26521)

Fixes #24618
FW-153 #resolve

PR Close #26521

gregmagolan committed:
build: add comment to /integration/bazel/package.json (#26488)

PR Close #26488

alexeagle committed:
refactor(bazel): allow google3 overrides for external deps (#26488)

PR Close #26488

alexeagle committed:
ci: don't run google3 tests on bazel external files (#26488)

PR Close #26488

alexeagle committed:
style: run buildifier (#26488)

PR Close #26488

alxhub committed:
refactor(ivy): obviate the Bazel component of the ivy_switch (#26550)

Originally, the ivy_switch mechanism used Bazel genrules to conditionally
compile one TS file or another depending on whether ngc or ngtsc was the
selected compiler. This was done because we wanted to avoid importing
certain modules (and thus pulling them into the build) if Ivy was on or
off. This mechanism had a major drawback: ivy_switch became a bottleneck
in the import graph, as it both imports from many places in the codebase
and is imported by many modules in the codebase. This frequently resulted
in cyclic imports which caused issues both with TS and Closure compilation.

It turns out ngcc needs both code paths in the bundle to perform the switch
during its operation anyway, so import switching was later abandoned. This
means that there's no real reason why the ivy_switch mechanism needed to
operate at the Bazel level, and for the ivy_switch file to be a bottleneck.

This commit removes the Bazel-level ivy_switch mechanism, and introduces
an additional TypeScript transform in ngtsc (and the pass-through tsc
compiler used for testing JIT) to perform the same operation that ngcc
does, and flip the switch during ngtsc compilation. This allows the
ivy_switch file to be removed, and the individual switches to be located
directly next to their consumers in the codebase, greatly mitigating the
circular import issues and making the mechanism much easier to use.

As part of this commit, the tag for marking switched variables was changed
from __PRE_NGCC__ to __PRE_R3__, since it's no longer just ngcc which
flips these tags. Most variables were renamed from R3_* to SWITCH_* as well,
since they're referenced mostly in render2 code.

Test strategy: existing test coverage is more than sufficient - if this
didn't work correctly it would break the hello world and todo apps.

PR Close #26550

alan-agius4 committed:
docs: fix links to browserlist (#26531)

PR Close #26531

marclaval committed:
feat(ivy): support providers and viewProviders (#25803)

PR Close #25803

robphoenix committed:
docs: update links to Karma homepage (#26509)

PR Close #26509

robphoenix committed:
docs: update http to https where possible (#26509)

PR Close #26509

gkalpak committed:
build(docs-infra): only render code example content in one place (#26514)

PR Close #26514

gkalpak committed:
build(docs-infra): allow `""` as empty region in `{@example}` tags (#26514)

PR Close #26514

gkalpak committed:
ci(docs-infra): move deployment to CircleCI (#26377)

PR Close #26377

gkalpak committed:
ci: fix `payload-size.sh` (#26377)

PR Close #26377

alan-agius4 committed:
build(docs-infra): update git ref for cli command docs (#26545)

This is to get the latest changes done in https://github.com/angular/angular-cli/pull/12634

PR Close #26545

gkalpak committed:
ci(docs-infra): remove jobs from Travis config (#26377)

PR Close #26377

mandarini committed:
docs: add Katerina Skroumpelou to GDE resources (#26547)

PR Close #26547

kara committed:
ci: add forms-overview guide to forms approval list (#26549)

PR Close #26549

gregmagolan committed:
build: review comments addressed (#26488)

PR Close #26488

mhevery committed:
release: cut the v7.0.0 release

mhevery committed:
docs: fix incomplete sentence in CHANGELOG.md

jenniferfell committed:
docs: reorganize events to highlight recent events before older events and move mix to past (#26551)

PR Close #26551

IgorMinar committed:
feat(docs-infra): add v6 to the aio version picker

petebacondarwin committed:
fix(docs-infra): update overload rendering (#24976)

Based on the review here:
https://github.com/angular/angular/pull/24976#issuecomment-415535125

PR Close #24976

mhevery committed:
docs: add release announcement URL

IgorMinar committed:
feat(docs-infra): update to @angular/* v7 (#26555)

PR Close #26555

jenniferfell committed:
docs: fix release info navigation (#26558)

PR Close #26558

matsko committed:
fix(ivy): ensure `ngClass` works with [class] bindings (#26559)

PR Close #26559

IgorMinar committed:
fix(docs-infra): correct the version of @angular-devkit/build-angular (#26555)

This dependency has not been correctly updated and I had to update it manually.

I think the issue is related to https://github.com/angular/angular-cli/issues/12624 which was fixed between CLI RCs and final.

PR Close #26555

fisherspy committed:
docs: add ant design mobile of angular in resources (#26562)

PR Close #26562

petebacondarwin committed:
build(docs-infra): move CLI API desciption higher up (#26568)

Closes #26556

PR Close #26568

petebacondarwin committed:
build: refactor ts-api-guardian jsdoc tag handling (#26595)

Allow the jsdoc tag processing to be configured by
type (export, member, param) and by action (required,
banned, toCopy).

This is a pre-requisite to moving over to using `@publicApi`
tags rather than `@stable` and `@experimental`.

PR Close #26595

petebacondarwin committed:
feat(docs-infra): remove stable & experimental from status selector (#26595)

PR Close #26595

petebacondarwin committed:
feat(docs-infra): rename tagdef: `@experimental` to `@publicApi` (#26595)

PR Close #26595

petebacondarwin committed:
docs: convert all `@experimental` tags to `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
build: update ts-api-guardian jsdoc tag requirements (#26595)

Now `@experimental` tags are banned; and `@publicApi`
tags are required on exports.

PR Close #26595

petebacondarwin committed:
docs: update platform-browser with `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update animations with `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: common with `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update core to use `@publicApi` tags (#26595)

PR Close #26595

matsko committed:
fix(ivy): ensure styling pipes are allocated before used in bindings (#26593)

PR Close #26593

petebacondarwin committed:
docs: update forms with `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update http with `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update platform-browser-dynamic to use `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update platform-server to use `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update platform-webworker to use `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update router to use `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update service-worker to use `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update upgrade to use `@publicApi` tags (#26595)

PR Close #26595

blove committed:
docs: add Brian Love to GDE resources (#26594)

PR Close #26594

petebacondarwin committed:
docs: update elements to use `@publicApi` tags (#26595)

PR Close #26595

petebacondarwin committed:
docs: update animations to use `@publicApi` tags (#26595)

PR Close #26595

kara committed:
build(ivy): fix benchmarks (#26601)

PR Close #26601

IgorMinar committed:
build: don't depend on npm packages from public api tests (#26602)

We need only the inputs from ng_module.

PR Close #26602

aciccarello committed:
docs(core): fix spelling in ComponentFactory (#26603)
PR Close #26603

kara committed:
fix(ivy): update largetable benchmark to use ngFor (#26604)

PR Close #26604

gregmagolan committed:
style: ran buildifier (#26488)

PR Close #26488

gregmagolan committed:
build: run CI (#26488)

PR Close #26488

mhevery committed:
build: remove manual tags from platform jasmine tests (#26606)

PR Close #26606

kara committed:
fix(ivy): update tree benchmark to use ngIf (#26608)

PR Close #26608

IgorMinar committed:
build: remove obsolete files (#26471)

PR Close #26471

blove committed:
docs: fix indentation (#26623)

PR Close #26623

gregmagolan committed:
build(bazel): fix //tools/testing:node target deps (#26629)

PR Close #26629

gregmagolan committed:
build: fix ivy package-dist generation (#26629)

PR Close #26629

gkalpak committed:
ci: run Xvfb in the background on CircleCI (#26596)

PR Close #26596

pkozlowski-opensource committed:
fix(ivy): support ViewContainerRef ng-container children (#26646)

Issue found while running NgPlural tests with ivy

PR Close #26646

gkalpak committed:
test(docs-infra): run basic smoke tests against PR previews (#26649)

This makes the tests run agaisnt the deployed production versions (as
part of the `aio_monitoring` job) more reliable.

PR Close #26649

gkalpak committed:
test(docs-infra): fix smoke tests (#26649)

PR Close #26649

gkalpak committed:
ci(docs-infra): notify caretaker about `aio_monitoring` failures (#26649)

PR Close #26649

LuisReinoso committed:
docs: add links to angular subdomains (#26653)

Add link to protactor.angular.io
Replace link from github.com/angular/universal to universal.angular.io
Replace link from github.com/angular/material2 to material.angular.io

fix #18257

PR Close #26653

gregmagolan committed:
build: add config_env_vars = ["compile"] to ngc-wrapped (#26471)

PR Close #26471

jenniferfell committed:
docs: update release info now that v7 is released (#26660)

PR Close #26660

IgorMinar committed:
build: fix the accept binary path in ts-api-guardian error message (#26602)

Fixes #26589

PR Close #26602

IgorMinar committed:
build: remove unused @angular/compiler ts-api-guardian golden files (#26602)

these files are not used because the api surface of this package is not considered
to be stable/public api.

the presence of these files only confuses the reader.

PR Close #26602

alexeagle committed:
build: Move non-bazel deps to devDependencies (#26691)

This makes yarn_install of ngdeps under Bazel faster, since we don't need many of the large dependencies.
It's important because downstream angular/bazel users will observe the same install time.

PR Close #26691

IgorMinar committed:
build: add aio/node_modules to the .bazelignore file (#26471)

PR Close #26471

matsko committed:
fix(ivy): ensure pipes indices are referenced in styling bindings (#26755)

PR Close #26755

IgorMinar committed:
build: add source-map-support to various rules within our repo (#26471)

This removes an verbose warning and adds source-mapping support to our build/ci.

Related issue: https://github.com/bazelbuild/rules_nodejs/issues/389

PR Close #26471

IgorMinar committed:
ci: build-packages-dist.sh should publish only regular and --compile=local packages (#26471)

We don't need to publish JIT compiled packages as this is not useful for real-world use-cases.

PR Close #26471

petebacondarwin committed:
build(docs-infra): mark code blocks to disable auto-linking (#26675)

You can now mark `<code>` blocks with a `no-auto-link` css class
to tell the code auto-linker to ignore this code block.

PR Close #26675

petebacondarwin committed:
build(docs-infra): ensure that CLI code blocks are not auto-linked (#26675)

Some of the text in CLI API docs were being auto-linked to API
pages. This was not correct, and in fact these blocks should not
have any auto links to Angular API at all.

Closes #26570

PR Close #26675

petebacondarwin committed:
build(docs-infra): refactor generateKeywords processor (#26676)

PR Close #26676

petebacondarwin committed:
build(docs-infra): include inherited members in search index (#26676)

Closes #23800

PR Close #26676

gkalpak committed:
ci: add more info on how CircleCI env vars are defined (#26596)

PR Close #26596

gkalpak committed:
ci: move CircleCI env variables definition to shell script (#26596)

PR Close #26596

alxhub commented:
@sculove this PR has test failures that seem directly related. Please address them and then we can merge the PR. Thanks!

IgorMinar committed:
build: add test-ivy-jit, test-fixme-ivy-jit, test-ivy-aot and test-fixme-ivy-aot yarn scripts (#26686)

PR Close #26686

IgorMinar committed:
build: rename the ivy compile mode 'local' to 'aot' (#26686)

PR Close #26686

petebacondarwin committed:
build(docs-infra): fix parameter type rendering (#26688)

Closes #24355

PR Close #26688

gkalpak committed:
test(docs-infra): make smoke tests more robust (#26649)

PR Close #26649

alexeagle committed:
ci: re-encrypt github publish token with recent ngcontainer (#26692)

Something about the docker image changed such that the encrypted value is different

PR Close #26692

IgorMinar committed:
ci: fix broken master branch due to conflicting merges (#26697)

PR Close #26697

alxhub committed:
build: fix BUILD file formatting

alexeagle committed:
ci: re-encrypt .circleci/github_token (#26698)

PR Close #26698

kara committed:
feat(ivy): support injecting the injector (#26699)

PR Close #26699

IgorMinar committed:
ci: don't publish ivy build artifacts on a non-master branch

matsko committed:
feat(ivy): expose a series of debug console tools (#26705)

PR Close #26705

gregmagolan committed:
fix(bazel): support --nolegacy_external_runfiles in protractor rule (#26708)

PR Close #26708

petebacondarwin committed:
build: update `canonical-path` dependency (#26719)

This new version (1.0.0) provides a typings file!

PR Close #26719

gkalpak committed:
ci: only publish builds if relevant `aio` jobs pass (#26722)

Some of the `aio`-/`docs`-related jobs rely on the locally built Angular
packages. When these jobs fail, it could mean that there is an issue
with the Angular packages (e.g. an unintentional breaking change).

This commit ensures that the `publish_artifacts` job is not run, unless
those `aio`-/`docs`-related jobs pass.

(The `test_aio_tools` job also uses the locally built Angular packages,
but it does not exercise them in a meaningful way to be worth making it
a prerequisite for `publish_artifacts`.)

PR Close #26722

gkalpak committed:
test(elements): make e2e tests for elements docs examples even less flaky (#26726)

PR Close #26726

ocombe committed:
docs: rename ivy compile mode 'local' to 'aot' in BAZEL.md (#26735)
PR Close #26735

alxhub committed:
fix(ivy): rename broken reflect_metadata switch file

This file was missed during the rename of the --define=compile
flag value from 'local' to 'aot. It's referenced as
reflect_metadata_${compile}.

gkalpak committed:
test(upgrade): make e2e tests for upgrade docs examples less flaky (#26726)

PR Close #26726

alxhub committed:
build(ivy): properly tag the js_expected_symbol_test rules

Previously the rules were tagged ivy-only, but missing the ivy-aot tag
which would cause them to run under CI.

alxhub committed:
build(ivy): fix typo in build file

gregmagolan committed:
style: format (#26708)

PR Close #26708

alxhub committed:
build: downgrade yarn.lock file to bazel generated version

gkalpak committed:
ci(docs-infra): fix deployment script (#26731)

PR Close #26731

alxhub committed:
docs: release notes for the v7.0.1 release

alxhub committed:
release: cut the v7.1.0-beta.0 release

mhevery committed:
fix(core): Remove static dependency from @angular/core to @angular/compiler (#26734)
PR Close #26734

IgorMinar committed:
build: remove ivy build/test tag hackery now that we can (#26735)

With https://github.com/bazelbuild/rules_nodejs/pull/388 fixed we can stop messing around
with tags and just rely on the explicitly defined tags.

PR Close #26735

IgorMinar committed:
ci: use yarn test-ivy-... (#26735)

This means that we use the same command to run test on CI as when developing locally.

PR Close #26735

petebacondarwin committed:
ci: exclude ngcc from g3sync check (#26740)

PR Close #26740

gkalpak committed:
build(docs-infra): upgrade cli command docs sources to 4faa81e25 (#26741)

PR Close #26741

gkalpak committed:
ci(docs-infra): fix deployment to Firebase (#26746)

Previously, `firebase-tools@3.x` was used and the deployment from
CircleCI failed with `Unexpected error` (HTTP code: 410).

This commit ensures that we use a recent version of `firebase-tools` for
deploying to Firebase. It also ensures that we use the locally installed
`firebase-tools` (not sure where it came from before :grin:).

PR Close #26746

gkalpak committed:
ci(docs-infra): reduce verbosity of `yarn build` on CI (#26746)

PR Close #26746

gkalpak committed:
build: clean up `*.gz` files created by `payload-size.sh` (#26746)

These files are not needed once the size has been calculated and there
is no point in keeping them around.

Deleting them prevents, for example, uploading unnecessary files from
`aio/dist/` to Firebase (because `deploy-to-firebase.sh` runs the
payload size checks right before deploying).

PR Close #26746

gkalpak committed:
ci(docs-infra): split `test_and_deploy_aio` to two separate jobs (#26746)

By splitting the jobs, if something goes wrong with deploying (e.g. a
network issue), we can re-run just that part instead of having to wait
for all the tests to complete again.

In terms of total duration, the difference should be minimal, because
the two operations (testing and deploying) do not depend on shared
tasks. For example, we need to build again (for the specific target
environment; e.g. stable, next, etc.) before deploying anyway.

PR Close #26746

manishbansal8843 committed:
docs: Removing duplicate lines (#26754)

Duplicate lines are removed to avoid confusion.
PR Close #26754

alxhub committed:
fix(compiler): generate relative paths only in summary file errors (#26759)

Previously errors in the summary file would include absolute file names.
This is undesirable as the output of a build should not depend on the
current working directory. Doing so causes nondeterminism issues in
builds.

PR Close #26759

matsko committed:
fix(ivy): ensure pipe declarations are populated lazily when a forward ref is detected (#26765)

PR Close #26765

kara committed:
fix(ivy): injecting optional TemplateRef on element should not throw (#26664)

PR Close #26664

kara committed:
fix(ivy): optional dependencies should not throw with module injector (#26763)

PR Close #26763

kara committed:
fix(ivy): resolve forward refs in providers (#26766)

PR Close #26766

scizor666 committed:
docs: replace unused variable by _ (#26768)

replace unused variable by _  to keep the code consistent between examples

PR Close #26768

mhevery committed:
build: fix comment formatting to make closure happy (#26769)
PR Close #26769

IgorMinar committed:
ci: configure chrome options for protractor (#26735)

so that they are more suitable for CI usage.

PR Close #26735

kara committed:
fix(ivy): host bindings should work on nodes with providers (#26771)

PR Close #26771

IgorMinar committed:
ci: mark //modules/benchmarks/src/largetable/render3:perf with fixme-ivy-* tags (#26735)

This target fails with odd error that we need to investigate:

[01:19:32] I/direct - Using ChromeDriver directly...
[01:19:32] E/direct - Error code: 135
[01:19:32] E/direct - Error message: Could not find chromedriver at /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/external/ngdeps/node_modules/webdriver-manager/selenium/chromedriver_2.41. Run 'webdriver-manager update' to download binaries.
[01:19:32] E/direct - Error: Could not find chromedriver at /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/external/ngdeps/node_modules/webdriver-manager/selenium/chromedriver_2.41. Run 'webdriver-manager update' to download binaries.

PR Close #26735

pickfire committed:
docs: fix typo (#26773)
PR Close #26773

IgorMinar committed:
test(ivy): fix or disable failing ivy tests (#26735)

These tests were previously not running on CI so they have always been broken,
or got broken just recently :-(.

test(ivy): mark failing test targets with fixme-ivy-jit and fixme-ivy-local tags

PR Close #26735

IgorMinar committed:
build: fix symbol-extractor by passing through "compile" as environmental variable (#26735)

PR Close #26735

kara committed:
fix(compiler): generate inputs with aliases properly (#26774)

PR Close #26774

jiangying000 committed:
docs: fix typo (#26776)
PR Close #26776

petebacondarwin committed:
test(core): fix security example e2e test (#26777)

The changes in d5cbcef0ea8e3c caused this test to fail.

PR Close #26777

petebacondarwin committed:
docs(core): update security guide sanitize example (#26777)

The sanitizer now removes the content of script tags as well as the tag itself.

PR Close #26777

danday74 committed:
docs: describe your change... (#26784)
PR Close #26784

matsko committed:
build: fix the golden files for hello_world_r2

matsko committed:
build: fix formatting issue in host_binding_spec file

matsko committed:
fix(ivy): ensure falsy styling is not applied during creation mode (#26793)

PR Close #26793

kara committed:
fix(ivy): host injection flag should not throw for embedded views (#26795)

PR Close #26795

devversion committed:
build: use bazel version from node modules (#26691)

* No longer depends on a custom CircleCI docker image that comes with Bazel pre-installed. Since Bazel is now available through NPM, we should be able to use the version from `@bazel/bazel` in order to enforce a consistent environment on CI and locally.
* This also reduces the amount of packages that need to be published (ngcontainer is removed)

PR Close #26691

kara committed:
fix(ivy): host bindings after dirs without host bindings should work (#26801)

PR Close #26801

kara committed:
fix(ivy): init hooks should be called before host bindings (#26802)

PR Close #26802

IgorMinar committed:
test: mark public_api_guard test with fixme-ivy-aot and no-ivy-jit (#26602)

These tests are currently broken because of the following reasons:
- ivy no longer emits a generated index, so the filename of the main d.ts file is different
- ivy currently exports some symbols that don't match the golden file, this needs investigation

PR Close #26602

devversion committed:
build: add option to disable ts-api-guardian default angular tag rules (#26761)

Since the API guardian can be also used by other projects, we should not set up the default Angular project tag rules unless specified explicitly through a given command option (`useAngularTagRules`)

PR Close #26761

pkozlowski-opensource committed:
fix(ivy): compile queries in JIT mode (#26816)

PR Close #26816

gkalpak committed:
test(docs-infra): make redirection tests more robust (#26649)

PR Close #26649

brandonroberts committed:
docs: update example dependencies to Angular/CLI/Universal V7 releases (#26820)

PR Close #26820

gregmagolan committed:
build(bazel): generalize fix for AMD name for flat module out file (#26837)

PR Close #26837

devversion committed:
build: fix ts-api-guardian does not work on windows w/ bazel (#26761)

* Fixes that the `ts-api-guardian` package does not work on Windows with Bazel. This is because `ts-api-guardian` does not resolve the runfiles through theNodeJS `require` function that properly handles runfiles within Bazel.

PR Close #26761

alxhub committed:
refactor(ivy): use wrapped metadata in all DecoratorHandlers (#26860)

Previously, the Directive, Injectable, and Pipe DecoratorHandlers were
directly returning @angular/compiler metadata from their analyze() steps.
This precludes returning any additional information along with that
metadata. This commit introduces a wrapper interface for these handlers,
opening the door for additional information to be returned from analyze().

Testing strategy: this is a refactor commit, existing test coverage is
sufficient.

PR Close #26860

alxhub committed:
feat(compiler): ability to mark an InvokeFunctionExpr as pure (#26860)

Uglify and other tree-shakers attempt to determine if the invocation
of a function is side-effectful, and remove it if so (and the result
is unused). A /*@__PURE__*/ annotation on the call site can be used
to hint to the optimizer that the invocation has no side effects and
is safe to tree-shake away.

This commit adds a 'pure' flag to the output AST function call node,
which can be used to signal to downstream emitters that a pure
annotation should be added. It also modifies ngtsc's emitter to
emit an Uglify pure annotation when this flag is set.

Testing strategy: this will be tested via its consumers, by asserting
that pure functions are translated with the correct comment.

PR Close #26860

alxhub committed:
feat(ivy): setClassMetadata() for assigning decorator metadata (#26860)

This commit introduces the setClassMetadata() private function, which
adds metadata to a type in a way that can be accessed via Angular's
ReflectionCapabilities. Currently, it writes to static fields as if
the metadata being added was downleveled from decorators by tsickle.

The plan is for ngtsc to emit code which calls this function, passing
metadata on to the runtime for testing purposes. Calls to this function
would then be tree-shaken away for production bundles.

Testing strategy: proper operation of this function will be an integral
part of TestBed metadata overriding. Angular core tests will fail if this
is broken.

PR Close #26860

alxhub committed:
feat(ivy): capture the identifier of a decorator during reflection (#26860)

Previously the ReflectionHost API only returned the names of decorators
and not a reference to their TypeScript Identifier. This commit adds
the identifier itself, so that a consumer can write references to the
decorator.

Testing strategy: this commit is trivial, and the functionality will be
exercised by downstream tests.

PR Close #26860

alxhub committed:
feat(ivy): generator of setClassMetadata statements for Angular types (#26860)

This commit introduces generateSetClassMetadataCall(), an API in ngtsc
for generating calls to setClassMetadata() for a given declaration. The
reflection API is used to enumerate Angular decorators on the declaration,
which are converted to a format that ReflectionCapabilities can understand.
The reflection metadata is then patched onto the declared type via a call
to setClassMetadata().

This is simply a utility, a future commit invokes this utility for
each DecoratorHandler.

Testing strategy: tests are included which exercise generateSetClassMetadata
in isolation.

PR Close #26860

alxhub committed:
feat(ivy): emit metadata along with all Angular types (#26860)

This commit causes a call to setClassMetadata() to be emitted for every
type being compiled by ngtsc (every Angular type). With this metadata,
the TestBed should be able to recompile these classes when overriding
decorator information.

Testing strategy: Tests in the previous commit for
generateSetClassMetadataCall() verify that the metadata as generated is
correct. This commit enables the generation for each DecoratorHandler,
and a test is added to ngtsc_spec to verify all decorated types have
metadata generated for them.

PR Close #26860

kara committed:
fix(ivy): bindings should be checked in components created dynamically by root component' (#26864)

PR Close #26864

matsko committed:
docs: fix aot typo

kara committed:
fix(ivy): dynamically created components should run init hooks (#26864)

PR Close #26864

alan-agius4 committed:
fix(bazel): unknown replay compiler error in windows (#26711)

In Windows the compiler path ends with `.exe` thus it will never match and throw  `Unknown replay compiler`

PR Close #26711

mhevery committed:
docs: add `Function Inlining` rules to `PERF_NOTES.md` (#26876)

PR Close #26876

kara committed:
fix(ivy): sync exported r3 symbol with compiler generated symbol (#26877)

PR Close #26877

matsko committed:
build: remove problematic `yarn` prefix for running bazel

matsko committed:
docs: release notes for the v7.0.2 release

matsko committed:
release: cut the v7.1.0-beta.1 release

hariberry committed:
docs: Webcomponents activated as of firefox 63 (#26889)
PR Close #26889

petebacondarwin committed:
style(router): remove trailing whitespace (#26891)

PR Close #26891

IgorMinar committed:
test: fix broken golden file for packages/core/test/bundling/hello_world_r2 (#26893)

PR Close #26893

sculove committed:
test(platform-browser): fix test codes for hammer_gestures

sculove commented:
ref https://github.com/angular/angular/pull/26925
